### PR TITLE
Docker Create Dockerfiles for IVA, Tomcat, Daemon, Init, OpenCGA CLI 

### PR DESCRIPTION
**Pull request by marrobi**
DO NOT MERGE

As per issue #949 @jjcollinge to take over.

### CODE REVIEW

**Comment by j-coll** [comment]
I would like to avoid, as much as possible, to copy the configuration files in `opencga-app/app/scripts/docker/opencga-init/conf/`, as they may change in the future.

We can add an extra copy step in the `opencga-app/pom.xml` file, so the generated `build` folder contains everything as expected.
See https://github.com/opencb/opencga/blob/develop/opencga-app/pom.xml#L232-L322

Similar with the IVA. I think that the best option would be to download the latest tar.gz from https://github.com/opencb/iva/releases

**Comment by marrobi** [comment]
@j-coll as long as we know where to seed the config from not a problem. Best to discuss with @jjcollinge .

**Comment by jjcollinge** [comment]
volumes are currently mounted at `/opt/opencga/conf` - should change to mount the volume at `/opt/volume` which will have a `conf` and `sessions` directory. We will then create a symlink from `/opt/opencga/conf` to `/opt/volume/conf` and `/opt/opencga/sessions` to `/opt/volume/sessions`.

File: opencga-app/app/scripts/docker/opencga-init/init-config.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+# Inject search and clinical hosts
+storage_config["search"]["host"] = args.search_host
+storage_config["clinical"]["host"] = args.clinical_host
+
+# Load configuration yaml
+with open(args.config_path) as f:
+    config = yaml.load(f)
+
+# Inject catalog database
+config["catalog"]["database"]["hosts"][0] = args.catalog_database_host
```
Comment (jjcollinge) [inline opencga-app/app/scripts/docker/opencga-init/init-config.py]: Add support for multiple hosts

Comment (jjcollinge) [reply opencga-app/app/scripts/docker/opencga-init/init-config.py]: Done in commit (https://github.com/opencb/opencga/pull/969/commits/166f39594a0f7d55faf09583c087fd4ead965ba5)

File: opencga-app/app/scripts/docker/opencga-init/setup.sh [hunk 2]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+
+# TODO
+# Change hbase.client.keyvalue.maxsize to 0 in /etc/hbase/conf/hbase-site.xml on hadoop cluster to avoid 'KeyValue size too large' error
+# by sshing to your (hdinsight cluster e.g. ssh sshuser@bartcga-ssh.azurehdinsight.net) and changing the xml (e.g. vi /etc/hbase/conf/hbase-site.xml)
```
Comment (lawrencegripper) [inline opencga-app/app/scripts/docker/opencga-init/setup.sh]: @bart-jansen is this something we need to do before we'll be able to use HBase correctly, I seem to remember it was but I might be wrong.

File: opencga-app/app/scripts/docker/opencga-app/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+tar xzvf /tmp/tomcat.tgz -C /opt && \
+mv /opt/apache-tomcat-${TOMCAT_VERSION} /opt/tomcat && \
+rm /tmp/tomcat.tgz && \
+rm -rf /opt/tomcat/webapps/examples && \
+rm -rf /opt/tomcat/webapps/docs && \
+rm -rf /opt/tomcat/webapps/ROOT && \
+chown -R 1001:1001 /opt/tomcat/
+
+# copy opencga build to tomcat server  - permissions??
+RUN cp /opt/opencga/opencga.war /opt/tomcat/webapps/opencga.war &&  chmod 777 /opt/opencga/
```
Comment (lawrencegripper) [inline opencga-app/app/scripts/docker/opencga-app/Dockerfile]: Is this `permissions??` something we need to look at?

File: opencga-app/app/scripts/docker/opencga-init/test/test-conf.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+assert(config["catalog"]["database"]["hosts"][2] == "test-catalog-database-host3")
+assert(config["catalog"]["database"]["user"] == "test-catalog-database-user")
+assert(config["catalog"]["database"]["password"] == "test-catalog-database-password")
+assert(config["catalog"]["database"]["options"]["enableSSL"] == True)
+assert(config["catalog"]["search"]["hosts"][0] == "test-catalog-search-host1")
+assert(config["catalog"]["search"]["hosts"][1] == "test-catalog-search-host2")
+assert(config["catalog"]["search"]["user"] == "test-catalog-search-user")
+assert(config["catalog"]["search"]["password"] == "test-catalog-search-password")
+assert(client_config["rest"]["host"] == "test-rest-host")
+assert(client_config["grpc"]["host"] == "test-grpc-host")
```
Comment (lawrencegripper) [inline opencga-app/app/scripts/docker/opencga-init/test/test-conf.py]: Nice, cool having the tests here!

File: opencga-app/app/scripts/docker/opencga-java-base/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# up version?
+FROM ubuntu:16.04
```
Comment (lawrencegripper) [inline opencga-app/app/scripts/docker/opencga-java-base/Dockerfile]: The webhosts are 18.04-LTS so probably makes sense to bump this up a version for consistency

File: opencga-app/app/scripts/docker/opencga/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+FROM opencga-build as build
+FROM opencga-java-base
+
+# Get opencga
+WORKDIR /opt/opencga
+
+COPY --from=build /build/build /opt/opencga
+
+# from repo
+#RUN   wget --quiet ${OPENCGA_RELEASE_URL} && mkdir /opt/opencga && tar xvf opencga-${OPENCGA_VERSION}.tar.gz --strip-components=1 -C /opt/opencga  && rm ./opencga-${OPENCGA_VERSION}.tar.gz
```
Comment (lawrencegripper) [inline opencga-app/app/scripts/docker/opencga/Dockerfile]: Nit: Remove unused code

File: opencga-app/app/scripts/docker/opencga-init/setup.sh (hunk 2 above)
Comment (bart-jansen) [reply opencga-app/app/scripts/docker/opencga-init/setup.sh]: @lawrencegripper Correct, at least before you index large loads using opencga. 

This one-liner should do the trick: 
```sudo sed -i '/<name>hbase.client.keyvalue.maxsize<\/name>/!b;n;c<value>0</value>' /etc/hbase/conf/hbase-site.xml```

File: opencga-app/app/scripts/docker/opencga-app/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,30 @@
+FROM opencga
+
+# Get Tomcat
+ARG TOMCAT_VERSION="8.5.35"
```
Comment (lawrencegripper) [inline opencga-app/app/scripts/docker/opencga-app/Dockerfile]: Maybe should add some of the schema org labels - http://label-schema.org/rc1/

File: opencga-app/app/scripts/docker/opencga/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,12 @@
+
+FROM opencga-build as build
+FROM opencga-java-base
+
+# Get opencga
+WORKDIR /opt/opencga
+
+COPY --from=build /build/build /opt/opencga
+
+# from repo
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply opencga-app/app/scripts/docker/opencga/Dockerfile]: I think using the build for now gives us more flexibility when testing then once we've stabilized we'd move to using the GH releases. What do you think?

[the remark this answers is not in the document]
Comment (marrobi) [reply opencga-app/app/scripts/docker/opencga/Dockerfile]: Agree, although discussed with @jjcollinge who  pointed out the images should already exist as an output of CI/CD, so these are unlikely to be built be "end users", only when doing dev/test. So may be best to leave with a build container? Either works - think getting CI/CD pushing some images to Docker hub for releases and maybe develop branch would be useful moving forward?

**Comment by jjcollinge** [comment]
Latest images to use:
- dotjson/opencga-app:azure
- dotjson/iva:test
- dotjson/opencga-init:azure
- dotjson/opencga-daemon:azure

**Pull request merged**